code/webauthn/contracts/contracts/Account.sol

```diff
@@ -0,0 +1,336 @@
+    _transaction.processPaymasterInput();
+  }
+
+  // ANCHOR: updateR1Owner
+  function updateR1Owner(bytes memory _r1Owner) external onlySelf {
+    r1Owner = _r1Owner;
+  }
+  // ANCHOR_END: updateR1Owner
+
+  // ANCHOR: validateWebAuthnSignature
+  function validateWebAuthnSignature(bytes memory webauthnSignature, bytes32 txHash) private view returns (bool valid) {
+    if (r1Owner.length == 0) {
+      return false;
+    }
+    bytes32[2] memory pubKey = abi.decode(r1Owner, (bytes32[2]));
+    valid = _validateWebAuthnSignature(txHash, webauthnSignature, pubKey);
+  }
+  // ANCHOR_END: validateWebAuthnSignature
+
+  // ANCHOR: _validateWebAuthnSignature
+  function _validateWebAuthnSignature(
+    bytes32 txHash,
+    bytes memory webauthnSignature,
+    bytes32[2] memory pubKey
+  ) public view returns (bool valid) {
+    (bytes memory authenticatorData, bytes memory clientData, bytes32[2] memory rs) = _decodeWebAuthnSignature(
+      webauthnSignature
+    );
+
+    // challenge is a base64url
+    string memory challenge = extractChallengeFromClientData(clientData);
+    string memory encodedHash = txHashToBase64Url(txHash);
+
+    if (keccak256(abi.encodePacked(challenge)) != keccak256(abi.encodePacked(encodedHash))) {
+      return false;
+    }
+
+    // malleability check
+    if (rs[1] > lowSmax) {
+      return false;
+    }
+
+    // check if the webauthn user flags are set
+    if (authenticatorData[32] & AUTH_DATA_MASK != AUTH_DATA_MASK) {
+        return false;
+    }
+
+    bytes32 clientDataHash = sha256(clientData);
+    bytes32 message = sha256(bytes.concat(authenticatorData, clientDataHash));
+
+    valid = callVerifier(message, rs, pubKey);
+  }
+  // ANCHOR_END: _validateWebAuthnSignature
+
+  function _decodeWebAuthnSignature(
+    bytes memory webauthnSignature
+  ) private pure returns (bytes memory authenticatorData, bytes memory clientData, bytes32[2] memory rs) {
+    (authenticatorData, clientData, rs) = abi.decode(webauthnSignature, (bytes, bytes, bytes32[2]));
+  }
+
+  // ANCHOR: callVerifier
+  function callVerifier(bytes32 hash, bytes32[2] memory rs, bytes32[2] memory pubKey) internal view returns (bool) {
+    /**
+     * Prepare the input format
+     * input[  0: 32] = signed data hash
+     * input[ 32: 64] = signature r
+     * input[ 64: 96] = signature s
+     * input[ 96:128] = public key x
+     * input[128:160] = public key y
+     */
+    bytes memory input = abi.encodePacked(hash, rs[0], rs[1], pubKey[0], pubKey[1]);
+
+    (bool __, bytes memory output) = P256.staticcall(input);
+
+    // the precompiled contract does not return a false value
+    if (output.length == 0) {
+      return false;
+    }
+
+    return abi.decode(output, (bool));
+  }
+  // ANCHOR_END: callVerifier
+
+  function extractChallengeFromClientData(bytes memory clientDataJSON) public pure returns (string memory) {
+    bytes memory challengeSlice = slice(clientDataJSON, 36, 58);
```

Comment:
It seems like a fixed length for this particular app & implementation, but I'm not sure if it would work in every use case. I will add a note about that.

Reply:
I did some testing myself and the prefix does seem pretty stable:
```
{"type":"webauthn.get","challenge":"lfhnIt2cSNmQr6IatgSTnkLYM1kSE21XPDoOyp7ZqD0","origin":"http://localhost:5173","crossOrigin":false}
```

The suffix is randomized, but that's not being checked here: https://chromium.googlesource.com/chromium/src/+/master/content/browser/webauth/client_data_json.md